syntax = "proto2";  // using proto2 to specify default-values
       // could do proto3 but then our code gets more responsibility
       // consider stronger support for json in proto3

package experimentserver;
import "google/protobuf/timestamp.proto";

enum Status {
    UNKNOWN   = 0;
    STARTING  = 1;
    RUNNING   = 2;
    DONE      = 3;
    FAILED    = 4;
    CANCELLED = 5;
}

enum JobStatus {
    JOB_START = 0;
    JOB_RUNNING = 1;
    JOB_SUCCESS = 2;
    JOB_FAILURE = 3;
}


message DispatcherMessage {
	oneof object {
		Experiment experiment   = 1;
		Ensemble ensemble       = 2;
		Realization realization = 3;
		Step step               = 4;
		Job job                 = 5;
	}
}

message ExperimentId {
  required string id = 1 [default = "experiment"];
}
message EnsembleId {
  required ExperimentId experiment = 1;
  required string id         = 2 [default = "ensemble"];  // an ensemble is identified by a string, not index
}
message RealizationId {
  required EnsembleId ensemble = 1;
  required uint64 realization  = 2 [default = 0];
}
message StepId {
  required RealizationId realization = 1;
  required uint64 step               = 2 [default = 0]; // replace with string?
}
message JobId {
  required StepId step  = 1;
  required uint64 index = 2 [default = 0];
}

message Experiment {
  required ExperimentId id   = 1; // NOTE: in the message the id is optional since it often will be streamed
  optional Status status   = 2 [default = UNKNOWN];

  map<string, Ensemble> ensembles= 101;
}
message Ensemble {
  required EnsembleId id   = 1;
  optional Status status   = 2 [default = UNKNOWN];

  map<uint64, Realization> realizations= 101;
}
message Realization {
  required RealizationId id= 1;
  optional Status status   = 2 [default = UNKNOWN];
  optional bool active     = 3 [default = true];  // not derived - active/inactive by user or algorithm
  optional double start_time      = 4;  // can be derived from step-list
  optional double end_time        = 5;  // can be derived from step-list

  map<uint64, Step> steps= 101;
}
message Step {
  required StepId id         = 1;
  optional Status status   = 2 [default = UNKNOWN];
  optional double start_time = 3;  // can be derived from job-list
  optional double end_time   = 4;  // can be derived from job-list

  map<uint64, Job> jobs = 101;
}
message Job {
  required JobId id                             = 1;
  optional JobStatus status                     = 2 [default = JOB_START];
  optional google.protobuf.Timestamp start_time = 3;
  optional google.protobuf.Timestamp end_time   = 4;
  optional string name                          = 6;
  optional string error                         = 7;
  optional string stdout                        = 8;
  optional string stderr                        = 9;

  optional uint64 current_memory = 10;
  optional uint64 max_memory     = 11;
  optional int32 exit_code       = 12;
}

